src/H5MF.c

```diff
@@ -1154,15 +1167,21 @@ H5MF_xfree(H5F_t *f, H5FD_mem_t alloc_type, haddr_t addr, hsize_t size)
 
     /* If size of the freed section is larger than threshold, add it to the free space manager */
     if (size >= f->shared->fs_threshold) {
+        bool section_merged_or_shunk = false; /* Whether free space section was merged or shrunk away */
+
         assert(f->shared->fs_man[fs_type]);
 
 #ifdef H5MF_ALLOC_DEBUG_MORE
         fprintf(stderr, "%s: Before H5FS_sect_add()\n", __func__);
 #endif /* H5MF_ALLOC_DEBUG_MORE */
 
         /* Add to the free space for the file */
-        if (H5MF__add_sect(f, alloc_type, f->shared->fs_man[fs_type], node) < 0)
+        if (H5MF__add_sect(f, alloc_type, f->shared->fs_man[fs_type], node, &section_merged_or_shunk) < 0) {
+            if (section_merged_or_shunk)
+                node = NULL;
```

Comment:
Avoid double-freeing this section that was allocated above in case `H5MF__add_sect()` already freed it.